nucypher-core/src/dkg.rs

```diff
@@ -21,6 +109,280 @@ pub enum FerveoVariant {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "Invalid seed length")
+        }
+    }
+
+    /// This class handles keyring material for request keys, by allowing deterministic
+    /// derivation of `RequestSecretKey` objects based on labels.
+    #[derive(Clone, ZeroizeOnDrop, PartialEq)]
+    pub struct RequestKeyFactory(SecretBox<RequestKeyFactorySeed>);
+
+    impl RequestKeyFactory {
+        /// Creates a secret key factory using the given RNG.
+        pub fn random_with_rng(rng: &mut (impl CryptoRng + RngCore)) -> Self {
+            let mut bytes = SecretBox::new(RequestKeyFactorySeed::default());
+            rng.fill_bytes(bytes.as_mut_secret());
+            Self(bytes)
+        }
+
+        /// Creates a secret key factory using the default RNG.
+        pub fn random() -> Self {
+            Self::random_with_rng(&mut OsRng)
+        }
+
+        /// Returns the seed size required by
+        pub fn seed_size() -> usize {
+            SecretKeyFactorySeedSize::to_usize()
+        }
+
+        /// Creates a RequestKeyFactory factory using the given random bytes.
+        ///
+        /// **Warning:** make sure the given seed has been obtained
+        /// from a cryptographically secure source of randomness!
+        pub fn from_secure_randomness(
+            seed: &[u8],
+        ) -> Result<Self, InvalidRequestFactorySeedLengthError> {
+            if seed.len() != Self::seed_size() {
+                // TODO: should I just panic! here instead?
+                return Err(InvalidRequestFactorySeedLengthError);
+            }
+            Ok(Self(SecretBox::new(*RequestKeyFactorySeed::from_slice(
+                seed,
+            ))))
+        }
+
+        /// Creates an untyped bytestring deterministically from the given label.
+        /// This can be used externally to seed some kind of a secret key.
+        pub fn make_secret(
+            &self,
+            label: &[u8],
+        ) -> SecretBox<GenericArray<u8, RequestKeyFactoryDerivedKeySize>> {
+            let prefix = b"REQUEST_SECRET_DERIVATION/";
+            let info = [prefix, label].concat();
+            kdf::<RequestKeyFactoryDerivedKeySize>(self.0.as_secret(), None, Some(&info))
+        }
+
+        /// Creates a `RequestSecretKey` deterministically from the given label.
+        pub fn make_key(&self, label: &[u8]) -> RequestSecretKey {
+            let prefix = b"REQUEST_KEY_DERIVATION/";
+            let info = [prefix, label].concat();
+            let seed =
+                kdf::<RequestKeyFactoryDerivedKeySize>(self.0.as_secret(), None, Some(&info));
+            let rng =
+                ChaCha20Rng::from_seed(<[u8; 32]>::try_from(seed.as_secret().as_slice()).unwrap());
+            RequestSecretKey::random_from_rng(rng)
+        }
+
+        /// Creates a `RequestKeyFactory` deterministically from the given label.
+        pub fn make_factory(&self, label: &[u8]) -> Self {
```

Comment:
What's the use case for creating a factory from another factory? I can understand it as a static method (not sure what's the name for those in Rust), but not much as an object method. You could potentially create a chain of factories, but I don't see why.

Reply:
Yeah I wasn't sure if there was some additional entropy reasoning, but your point is well taken.

Perhaps I'll remove the `make_factory` method for the `SessionSecretFactory` here and we can revisit the Umbral Factory in a separate PR.

Reply:
For some added context... some of the motivation here may have been to accommodate derivation of umbral secret key factories from Ursula's mnemonic using an HKDF, that are themselves subsequently used to derive keys for DecryptingPower.

Reply:
For the bigger picture, this is what now happens on the `nucypher/nucypher` side (PR # 3134)  - https://github.com/derekpierre/nucypher/blob/25519/nucypher/crypto/keystore.py#L439. Unsure if that is what we want. 

Thoughts?